### DISCUSSION TITLE
Error handeling

### DISCUSSION THREAD

**Discussion by ShakedHad** [urql-graphql/urql#2637] [Q&A] [answered]
Hi All,

I'm trying to configure URQL client to throw an error when it receives a GraphQL error (so I can handle it using throw catch)

I've tried configuring my client like this:
```
const client = createClient({
  url: process.env.SERVER_URL,
  exchanges: [
    errorExchange({
      onError: (error: CombinedError) => {
        throw error;
      },
    }),
    ...defaultExchanges,
  ],
});
```

but when throwing the error in errorExchange it won't be catched in the try-catch surrounding the original client.query call.
Any idea what am i missing? What is the best practice to handle GraphQL errors using URQL client?

**Comment by kitten** [comment, marked as answer]
I'm assuming you're using the `urql` bindings with React, right?
The timing here isn't synchronous, so at the time you're throwing that doesn't mean that you're in the same synchronous stack as React's render. In fact, because you're throwing in the exchanges when you're receiving a response, you'll always basically be throwing "before" the response even reaches React, regardless of when the result comes in.

In short, in React today, if you'd like to handle all errors (or some) in your error boundaries, then in your render function / function component you'd basically want to add a line to rethrow the errors, e.g.

```js
const [result] = useQuery(...);

if (result.error) {
  throw result.error;
}
```

**Reply by ShakedHad** [reply]
Thank you for your answer!
I'm actually using urql in a node express server.
I'm creating and exporting the client once the server is app and using it in the controllers for queries and mutations

**Reply by kitten** [reply]
I see, in that case you'll basically have to chain the error handling, i.e.

```js
client.query(...).toPromise().then(result => {
  if (result.error) {
    throw result.error;
  } else {
    return result;
});
```

The same principle applies there basically, but the reason why this isn't the default is because GraphQL can deliver partial results on error and it isn't always desireable to flat-out fail by default, since it'd create more cases that'd need to then be handled and "reconverged" in user-code.

**Reply by ShakedHad** [reply]
Oh, I understand.
Thank you so much for your help
